libvast/src/system/active_partition.cpp

```diff
@@ -341,64 +335,63 @@ active_partition_actor::behavior_type active_partition(
-      self->state.data.synopsis.unshared().add(
-        x, self->state.partition_capacity, self->state.synopsis_index_config);
-      size_t col = 0;
-      for (const auto& [field, offset] :
-           caf::get<record_type>(schema).leaves()) {
-        const auto qf = qualified_record_field{schema, offset};
-        auto& idx = self->state.indexers[qf];
-        if (should_skip_index_creation(
-              field.type, qf, self->state.synopsis_index_config.rules)) {
-          continue;
-        }
-        if (!idx) {
+  auto make_stage = [&] {
+    return detail::attach_notifying_stream_stage(
+      self, false,
+      [=](caf::unit_t&) {
+        // nop
+      },
+      [=](caf::unit_t&, caf::downstream<table_slice>& out, table_slice x) {
+        VAST_TRACE_SCOPE("partition {} got table slice {} {}",
+                         self->state.data.id, VAST_ARG(out), VAST_ARG(x));
+        // The index already sets the correct offset for this slice, but in some
+        // unit tests we test this component separately, causing incoming table
+        // slices not to have an offset at all. We should fix the unit tests
+        // properly, but that takes time we did not want to spend when migrating
+        // to partition-local ids. -- DL
+        if (x.offset() == invalid_id)
+          x.offset(self->state.data.events);
+        VAST_ASSERT(x.offset() == self->state.data.events);
+        // Adjust the import time range iff necessary.
+        auto& mutable_synopsis = self->state.data.synopsis.unshared();
+        mutable_synopsis.min_import_time = std::min(
+          self->state.data.synopsis->min_import_time, x.import_time());
+        mutable_synopsis.max_import_time = std::max(
+          self->state.data.synopsis->max_import_time, x.import_time());
+        // We rely on `invalid_id` actually being the highest possible id
+        // when using `min()` below.
+        static_assert(invalid_id == std::numeric_limits<vast::id>::max());
+        auto first = x.offset();
+        auto last = x.offset() + x.rows();
+        const auto& schema = x.schema();
+        VAST_ASSERT(!schema.name().empty());
+        auto it = self->state.data.type_ids.emplace(schema.name(), ids{}).first;
+        auto& ids = it->second;
+        VAST_ASSERT(first >= ids.size());
+        // Mark the ids of this table slice for the current type.
+        ids.append_bits(false, first - ids.size());
+        ids.append_bits(true, last - first);
+        self->state.data.events += x.rows();
+        self->state.data.synopsis.unshared().add(
+          x, self->state.partition_capacity, self->state.synopsis_index_config);
+        size_t column_idx = -1;
+        for (const auto& [field, offset] :
+             caf::get<record_type>(schema).leaves()) {
+          column_idx++;
+          // TODO: The qualified record field is a leftover from heterogeneous
+          // partitions, the indexers can be indexed by the column instead.
```

Comment:
I think we should index them by offset instead of flat index, as we may want to create an index for a field in a record in a list, for example.